torch/csrc/utils/tensor_new.cpp

```diff
@@ -45,87 +45,58 @@ namespace torch { namespace utils {
+  maybe_initialize_cuda(options.device());
   pybind11::gil_scoped_release no_gil;
-  return torch::zeros(sizes, options(dispatch_key, scalar_type, device));
+  return torch::zeros(sizes, build_options(options, scalar_type, device));
 }
 
-Tensor dispatch_ones(c10::DispatchKey dispatch_key, at::ScalarType scalar_type, const optional<Device>& device, IntArrayRef sizes) {
-  maybe_initialize_cuda(dispatch_key);
+Tensor dispatch_ones(c10::TensorOptions options, at::ScalarType scalar_type, const optional<Device>& device, IntArrayRef sizes) {
+  maybe_initialize_cuda(options.device());
   pybind11::gil_scoped_release no_gil;
-  return torch::ones(sizes, options(dispatch_key, scalar_type, device));
+  return torch::ones(sizes, build_options(options, scalar_type, device));
 }
 
-Tensor dispatch_full(c10::DispatchKey dispatch_key, at::ScalarType scalar_type, Scalar fill_value, const optional<Device>& device, IntArrayRef sizes) {
-  maybe_initialize_cuda(dispatch_key);
+Tensor dispatch_full(c10::TensorOptions options, at::ScalarType scalar_type, Scalar fill_value, const optional<Device>& device, IntArrayRef sizes) {
+  maybe_initialize_cuda(options.device());
   pybind11::gil_scoped_release no_gil;
-  return torch::full(sizes, fill_value, options(dispatch_key, scalar_type, device));
+  return torch::full(sizes, fill_value, build_options(options, scalar_type, device));
 }
 
-Tensor new_with_sizes(c10::DispatchKey dispatch_key, at::ScalarType scalar_type, const optional<Device>& device, IntArrayRef sizes) {
-  maybe_initialize_cuda(dispatch_key);
+Tensor new_with_sizes(c10::TensorOptions options, at::ScalarType scalar_type, const optional<Device>& device, IntArrayRef sizes) {
+  maybe_initialize_cuda(options.device());
   pybind11::gil_scoped_release no_gil;
-  return torch::empty(sizes, options(dispatch_key, scalar_type, device));
+  return torch::empty(sizes, build_options(options, scalar_type, device));
 }
 
-Tensor new_with_storage(c10::DispatchKey dispatch_key, at::ScalarType scalar_type, Storage storage) {
-  auto tensor = at::empty({}, options(dispatch_key, scalar_type));
+Tensor new_with_storage(c10::TensorOptions options, at::ScalarType scalar_type, Storage storage) {
+  auto tensor = at::empty({}, build_options(options, scalar_type));
   tensor.set_(std::move(storage));
   return tensor;
 }
 
-Tensor new_with_tensor(c10::DispatchKey dispatch_key, at::ScalarType scalar_type, const Tensor& other) {
-  TORCH_CHECK_TYPE(legacyExtractDispatchKey(other.key_set()) == dispatch_key, "expected ",
-                   toString(dispatch_key), " (got ", toString(legacyExtractDispatchKey(other.key_set())), ")");
+Tensor new_with_tensor(c10::TensorOptions options, at::ScalarType scalar_type, const Tensor& other) {
+  TORCH_CHECK_TYPE(other.options().type_equal(options), "expected ",
+                   options, " (got ", other.options(), ")");
```

Comment:
This isn't right, because `other.options()` also includes dtype

Reply:
a) I'm not sure which thing isn't right here b) regardless, is it worth putting in a comment?

Reply:
This comment is out of date; the code is updated!